### PR TITLE
Adjust feature flag handling in parseFeatures.go and features.go

### PR DESCRIPTION
**Pull request by danqixu**
Issue: #3641

### CODE REVIEW

File: buildcontext/parsefeatures.go [hunk 1]
```diff
@@ -27,6 +28,13 @@ func parseFeatures(buildFilePath string, featureFlagOverrides string, projectRef
 		return nil, fmt.Errorf("No version specified in %s/Earthfile", projectRef)
 	}
 
+	warningStrs, err := ftrs.Adjust()
+	if err != nil {
+		return nil, err
+	}
+
+	console.Printf("Note that the feature flag %s is not necessary for version %s, since the feature was enabled by default. You may remove this flag from your Earthfile.", strings.Join(warningStrs, ", "), ftrs.Version())
```
Comment (alexcb) [inline buildcontext/parsefeatures.go:36]: this should only be printed when `warningStrs` has len > 0

File: features/features.go [hunk 2]
```diff
@@ -309,3 +255,68 @@ func FromContext(ctx context.Context) *Features {
...
+
+	if f.ArgScopeSet && !f.ShellOutAnywhere {
+		// ArgScopeSet uses new ARG declaration logic that requires
+		// ShellOutAnywhere. We're erroring here to ensure that users get that
+		// feedback as early as possible.
+		return nil, errors.New("--arg-scope-and-set requires --shell-out-anywhere")
+	}
+
+	return warningStrs, nil
+}
```
Comment (alexcb) [inline features/features.go:322]: nit: this is not a POSIX file; we need a final newline.

**Review by alexcb — COMMENTED** [review]
I like the approach, but is there a way we can avoid duplicating the `--flag-names`? maybe we could use reflection to pull them out of the `Features` struct?

**Comment by danqixu** [comment]
> I like the approach, but is there a way we can avoid duplicating the `--flag-names`? maybe we could use reflection to pull them out of the `Features` struct?


There is no good way. If reflection is used, we still need to fill in the serial numbers of all fields. Once the order in the structure changes, we still need to remember to modify this code, which is not intuitive and easy to maintain. I thought about using reflection but gave up

**Comment by alexcb** [comment]
> > I like the approach, but is there a way we can avoid duplicating the `--flag-names`? maybe we could use reflection to pull them out of the `Features` struct?
> 
> There is no good way. If reflection is used, we still need to fill in the serial numbers of all fields. Once the order in the structure changes, we still need to remember to modify this code, which is not intuitive and easy to maintain. I thought about using reflection but gave up

I wonder if we could use the Features struct's tags to also include the earthly version where the values are enabled by default?

If you haven't already, maybe look at github.com/jessevdk/go-flags for ideas of how it uses reflection to read the tags as well as set the fields in the struct.

**Comment by danqixu** [comment]
If you don't mind the problem I asked, I can use reflection

like:
```golang
// Enable version-specific features.
    if versionAtLeast(*f, 0, 5) {
        enableFunc(&f.ExecAfterParallel, 3)
        enableFunc(&f.ParallelLoad, 4)
        enableFunc(&f.UseRegistryForWithDocker, 5)
    }
    if versionAtLeast(*f, 0, 6) {
        enableFunc(&f.ForIn, 6)
        enableFunc(&f.NoImplicitIgnore, 7)
        enableFunc(&f.ReferencedSaveOnly, 8)
        enableFunc(&f.RequireForceForUnsafeSaves, 9)
        enableFunc(&f.UseCopyIncludePatterns, 10)
    }
...
```

File: features/features.go (hunk 2 above)
Comment (danqixu) [reply features/features.go]: amended

**Comment by alexcb** [comment]
> If you don't mind the problem I asked, I can use reflection
> 
> like:
> 
> ```go
> // Enable version-specific features.
>     if versionAtLeast(*f, 0, 5) {
>         enableFunc(&f.ExecAfterParallel, 3)
>         enableFunc(&f.ParallelLoad, 4)
>         enableFunc(&f.UseRegistryForWithDocker, 5)
>     }
>     if versionAtLeast(*f, 0, 6) {
>         enableFunc(&f.ForIn, 6)
>         enableFunc(&f.NoImplicitIgnore, 7)
>         enableFunc(&f.ReferencedSaveOnly, 8)
>         enableFunc(&f.RequireForceForUnsafeSaves, 9)
>         enableFunc(&f.UseCopyIncludePatterns, 10)
>     }
> ...
> ```

this is really hard to follow.

I think looking into using tags, as I previously suggested, would be cleaner.

**Comment by danqixu** [comment]
> > If you don't mind the problem I asked, I can use reflection
> > like:
> > ```go
> > // Enable version-specific features.
> >     if versionAtLeast(*f, 0, 5) {
> >         enableFunc(&f.ExecAfterParallel, 3)
> >         enableFunc(&f.ParallelLoad, 4)
> >         enableFunc(&f.UseRegistryForWithDocker, 5)
> >     }
> >     if versionAtLeast(*f, 0, 6) {
> >         enableFunc(&f.ForIn, 6)
> >         enableFunc(&f.NoImplicitIgnore, 7)
> >         enableFunc(&f.ReferencedSaveOnly, 8)
> >         enableFunc(&f.RequireForceForUnsafeSaves, 9)
> >         enableFunc(&f.UseCopyIncludePatterns, 10)
> >     }
> > ...
> > ```
> 
> this is really hard to follow.
> 
> I think looking into using tags, as I previously suggested, would be cleaner.

I've updated my previous comment to clarify my concerns about using reflection to get the tag.

**Comment by alexcb** [comment]
I think there's a disconnect in communicate here; I was suggesting we add a new filed, such as:

```
enabled_in_version: "0.7"
```

to the `Features` struct,

So for example, this field

```
CheckDuplicateImages     bool `long:"check-duplicate-images" description:"check for duplicate images during output"`
```

would be updated to

```
CheckDuplicateImages     bool `long:"check-duplicate-images" description:"check for duplicate images during output" enabled_in_version: "0.7"`
```

Then we would use reflect (in the same way that github.com/jessevdk/go-flags works).

Using serial numbers of fields is too difficult to maintain.

**Comment by danqixu** [comment]
> I think there's a disconnect in communication here; I was suggesting we add a new filed, such as:
> 
> ```
> enabled_in_version: "0.7"
> ```
> 
> to the `Features` struct,
> 
> So for example, this field
> 
> ```
> CheckDuplicateImages     bool `long:"check-duplicate-images" description:"check for duplicate images during output"`
> ```
> 
> would be updated to
> 
> ```
> CheckDuplicateImages     bool `long:"check-duplicate-images" description:"check for duplicate images during output" enabled_in_version: "0.7"`
> ```
> 
> Then we would use reflect (in the same way that github.com/jessevdk/go-flags works).
> 
> Using serial numbers of fields is too difficult to maintain.

Oh I got it, I will try your suggestion.

**Comment by danqixu** [comment]
I have made the changes as requested, could you please check if there are any issues?

**Comment by danqixu** [comment]
@alexcb

File: features/features.go [hunk 3]
```diff
@@ -309,3 +252,57 @@ func FromContext(ctx context.Context) *Features {
...
+
+func ParseVersion(version string) (int, int) {
```
Comment (alexcb) [inline features/features.go:291]: since this function panics, we should rename it to `must....` to make it more clear.

```suggestion
func MustParseVersion(version string) (int, int) {
```

File: buildcontext/parsefeatures.go [hunk 4]
```diff
@@ -27,6 +28,15 @@ func parseFeatures(buildFilePath string, featureFlagOverrides string, projectRef
 		return nil, fmt.Errorf("No version specified in %s/Earthfile", projectRef)
 	}
 
+	warningStrs, err := ftrs.Adjust()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(warningStrs) > 0 {
+		console.Printf("Note that the feature flag %s is not necessary for version %s, since the feature was enabled by default. You may remove this flag from your Earthfile.", strings.Join(warningStrs, ", "), ftrs.Version())
```
Comment (alexcb) [inline buildcontext/parsefeatures.go:37]: Let's reword this to be a bit shorter:
```suggestion
		console.Printf("NOTE: The %s feature is enabled by default under VERSION %s, and can be safely removed from the VERSION command", strings.Join(warningStrs, ", "), ftrs.Version())
```

File: features/features.go [hunk 5]
```diff
@@ -309,3 +252,57 @@ func FromContext(ctx context.Context) *Features {
 	}
 	return nil
 }
+
+func (f *Features) Adjust() ([]string, error) {
```
Comment (alexcb) [inline features/features.go:256]: I think having "flags" appear in this method name would make it a bit more readable

```suggestion
func (f *Features) processFlags() ([]string, error) {
```

**Review by alexcb — COMMENTED** [review]
This is looking really good! just a few minor suggestions, thanks for sticking with it!

File: features/features.go [hunk 6]
```diff
@@ -309,3 +252,57 @@ func FromContext(ctx context.Context) *Features {
...
+		if len(version) == 0 {
+			continue
+		}
+		majorVersion, minorVersion := ParseVersion(field.Tag.Get("enabled_in_version"))
+		if versionAtLeast(*f, majorVersion, minorVersion) && value.Kind() == reflect.Bool {
+			if value.Bool() {
+				tagName := field.Tag.Get("long")
+				warningStrs = append(warningStrs, fmt.Sprintf("--%s", strings.ToLower(tagName)))
+			}
+		value.SetBool(true)
```
Comment (alexcb) [inline features/features.go:275]: indentation issue (although it didn't cause `+lint` to fail which is odd)
```suggestion
			value.SetBool(true)
```

**Review by alexcb — APPROVED** [review]
